### DISCUSSION TITLE
How it works in distributed applications?

### DISCUSSION THREAD

**Discussion by fiftin** [Unleash/unleash#4559] [Q&A] [answered]
Hi,

I have distributed node.js application with 1 million users per month and 5+ servers. I want to use Unleash for Canary deployment.

And I have question, how it works for distributed apps? Is it store users in memory of client app or getting it from the Unleash server? Is it possible to get out of sync between nodes?

**Comment by ivarconr** [comment]
Hi @fiftin, 
Unleash performs the evaluations locally in your applications. It does not store the users, and the rules are distributed from the Unleash API and stored locally in the SDK running in your application. It means that it will get out of sync, given you do not change the rules at the same time.

**Comment by nunogois** [comment, marked as answer]
Hi @fiftin,

To add to @ivarconr's reply:

1. **How does Unleash work for distributed apps?**
Unleash is deterministic, meaning it will evaluate feature toggle configurations in the same way across all nodes, given the same input. The feature toggle configurations are pulled from the Unleash server to the individual clients running in your distributed setup.

2. **Is it storing users in the memory of the client app or is it getting it from the Unleash server?**
Unleash does not store user-specific information. It focuses on feature toggle configurations, which are stored on the Unleash server and synced to client apps. The evaluation of a feature toggle would be done on the client side based on these configurations.

3. **Is it possible to get out of sync between nodes?**
Since feature toggle configurations are fetched from your Unleash server by your client SDKs, there's a possibility of them going temporarily out-of-sync if you update the feature toggle configuration, since the client SDKs need to fetch the latest configuration for that feature toggle in order to evaluate them in the same way. This would only happen when the configuration is changed, as it can take some time before all nodes receive the updates. Eventually and inevitably they will reach consistency and this is considered OK for most cases.

I would also recommend checking out [Edge](https://github.com/Unleash/unleash-edge), since it may be an interesting tool for the use case that you're describing.
